File: Practicas/RobocodeLabo-1.10.0/RobocodeLabo/src/estrategias/EstrategiaAntiWalls.java
package estrategias;

import laboratorio.NicoustinRobot;
import robocode.WinEvent;

/**
 * Estrategia Anti-Walls: Counter específico para robots tipo Walls
 * Usa el patrón "Corner Ambush" - emboscadas en esquinas
 */
public class EstrategiaAntiWalls implements Estrategia {
    
    private NicoustinRobot robot;
    private int currentCorner = 0; // 0=arriba-izq, 1=arriba-der, 2=abajo-der, 3=abajo-izq
    private boolean inPosition = false;
    private int waitTurns = 0;
    private boolean targetDetected = false;
    
    @Override
    public void runB(NicoustinRobot robot) {
        this.robot = robot;
        
        // Colores distintivos para counter
        robot.setColors(6, 0, 14); // Rojo/Negro/Naranja - cazador
        
        System.out.println("🎯 ANTI-WALLS: Iniciando caza de robots perimetrales");
        
        // Detectar tamaño del campo para posicionamiento preciso
        FieldDetector.detectFieldSize(robot);
        System.out.println("🎯 ANTI-WALLS: Campo " + FieldDetector.getFieldWidth() + "x" + FieldDetector.getFieldHeight() + " detectado");
        
        while(true) {
            if (!inPosition) {
                // Ir a la esquina objetivo
                moveToCorner(currentCorner);
            } else {
                // Estamos en posición - vigilar y disparar
                cornerAmbush();
            }
        }
    }
    
    /**
     * Mueve el robot a la esquina especificada usando tamaños reales del campo
     */
    private void moveToCorner(int corner) {
        int fieldWidth = FieldDetector.getFieldWidth();
        int fieldHeight = FieldDetector.getFieldHeight();
        int cornerDistance = Math.max(fieldWidth, fieldHeight) / 2; // Distancia para llegar cerca de esquina
        
        switch(corner) {
            case 0: // Esquina superior izquierda
                robot.turnLeft((int)(315 - robot.heading)); // Noreste
                robot.ahead(cornerDistance);  // Distancia calculada
                robot.turnRight((int)(135 - robot.heading)); // Apuntar hacia perímetro
                break;
            case 1: // Esquina superior derecha  
                robot.turnLeft((int)(225 - robot.heading)); // Noroeste
                robot.ahead(cornerDistance);
                robot.turnRight((int)(225 - robot.heading)); // Apuntar hacia perímetro
                break;
            case 2: // Esquina inferior derecha
                robot.turnLeft((int)(135 - robot.heading)); // Suroeste
                robot.ahead(cornerDistance);
                robot.turnRight((int)(315 - robot.heading)); // Apuntar hacia perímetro
                break;
            case 3: // Esquina inferior izquierda
                robot.turnLeft((int)(45 - robot.heading));  // Sureste
                robot.ahead(cornerDistance);
                robot.turnRight((int)(45 - robot.heading));  // Apuntar hacia perímetro
                break;
        }
        
        inPosition = true;
        waitTurns = 0;
        System.out.println("🎯 ANTI-WALLS: En posición esquina " + corner + " - Esperando target");
    }
    
    /**
     * Lógica de emboscada desde la esquina
     */
    private void cornerAmbush() {
        waitTurns++;
        
        if (!targetDetected) {
            // Escanear área perimetral
            robot.turnGunLeft(10); // Escaneo lento y preciso
        }
        
        // Si llevamos mucho tiempo sin detectar target, cambiar esquina
        if (waitTurns > 50) {
            changeCorner();
        }
        
        // Pequeños ajustes de posición para mantener ventaja
        if (waitTurns % 20 == 0) {
            adjustPosition();
        }
    }
    
    /**
     * Cambiar a la siguiente esquina estratégicamente
     */
    private void changeCorner() {
        currentCorner = (currentCorner + 1) % 4;
        inPosition = false;
        targetDetected = false;
        System.out.println("🎯 ANTI-WALLS: Cambiando a esquina " + currentCorner);
    }
    
    /**
     * Ajustar posición para mantener ventaja táctica
     */
    private void adjustPosition() {
        // Movimiento mínimo para no perder posición ventajosa
        robot.ahead(20);
        robot.back(10);
    }
    
    @Override
    public void onScannedRobot() {
        double distance = robot.scannedDistance;
        double bearing = robot.scannedBearing;
        
        targetDetected = true;
        
        // Disparar inmediatamente - estamos en emboscada
        if (distance < 200) {
            // Disparo certero a corta distancia
            robot.turnGunTo(robot.scannedAngle);
            robot.fire(3); // Máxima potencia
            System.out.println("🎯 ANTI-WALLS: ¡Emboscada exitosa! Target a " + (int)distance + "px");
        } else if (distance < 400) {
            // Disparo de seguimiento
            robot.turnGunTo(robot.scannedAngle);
            robot.fire(2);
            System.out.println("🎯 ANTI-WALLS: Disparo de seguimiento a " + (int)distance + "px");
        }
        
        // Si el target está muy cerca, ajustar posición
        if (distance < 100) {
            // Retroceder manteniendo el ángulo de disparo
            robot.turnRight(robot.scannedBearing + 180); // Dar la espalda
            robot.ahead(50); // Retroceder
            robot.turnRight(robot.scannedBearing); // Volver a apuntar
        }
    }
    
    @Override
    public void onHitByBullet() {
        // Si nos están disparando, cambiar posición inmediatamente
        System.out.println("🎯 ANTI-WALLS: ¡Detectado! Cambiando posición");
        
        // Escape táctico perpendicular
        robot.turnRight(robot.hitByBulletBearing + 90);
        robot.ahead(100);
        
        // Cambiar a otra esquina
        changeCorner();
    }
    
    @Override
    public void onHitWall() {
        // Si chocamos con pared, ajustar posición
        robot.back(30);
        robot.turnRight(45);
        System.out.println("🎯 ANTI-WALLS: Ajuste de pared");
    }
    
    @Override
    public void setRobot(NicoustinRobot robot) {
        this.robot = robot;
    }
    
    @Override
    public String evalStrat() {
        // La evaluación la hace el robot, no la estrategia individual
        return null;
    }
    
    @Override
    public void onWin(WinEvent e) {
        System.out.println("🏆 ANTI-WALLS: ¡Victoria táctica! Counter exitoso");
        
        // Celebración tipo francotirador
        for(int i = 0; i < 4; i++) {
            robot.turnGunRight(90);
            robot.ahead(25);
        }
    }
}

File: Practicas/Robocode-Kotlin/src/estrategias/EstrategiaAntiWalls.kt
package estrategias

import laboratorio.NicoustinRobot
import robocode.WinEvent
import java.lang.System.out
import kotlin.math.*

class EstrategiaAntiWalls : Estrategia {

    private lateinit var robot: NicoustinRobot
    private var currentCorner: Int = 0
    private var inPosition: Boolean = false
    private var waitTurns: Int = 0
    private var targetDetected: Boolean = false

    override fun runB(robot: NicoustinRobot) {
        this.robot = robot

        // Seteamos colores
        robot.setColors(6, 0, 14)

        out.println("ANTI-WALLS: Iniciando caza de robots perimetrales")

        // Detectar tamaño del campo para posicionamiento preciso
        FieldDetector.detectFieldSize(robot)
        out.println("ANTI-WALLS: Campo '${FieldDetector.getFieldWidth()}' x '${FieldDetector.getFieldHeight()}' detectado")

        while (true) {
            if (!inPosition) {
                moveToCorner(currentCorner)
            } else {
                cornerAmbush()
            }
        }
    }

    private fun moveToCorner(corner: Int) {
        val fieldWidth: Int = FieldDetector.getFieldWidth()
        val fieldHeight: Int = FieldDetector.getFieldHeight()
        val cornerDistance = max(fieldWidth, fieldHeight) / 2

        when (corner) {
            0 -> {
                robot.turnLeft((315 - robot.heading).toInt())
                robot.ahead(cornerDistance)
                robot.turnRight((135 - robot.heading).toInt())
            }
            1 -> {
                robot.turnLeft((225 - robot.heading).toInt())
                robot.ahead(cornerDistance)
                robot.turnRight((225 - robot.heading).toInt())
            }
            2 -> {
                robot.turnLeft((135 - robot.heading).toInt())
                robot.ahead(cornerDistance)
                robot.turnRight((315 - robot.heading).toInt())
            }
            3 -> {
                robot.turnLeft((45 - robot.heading).toInt())
                robot.ahead(cornerDistance)
                robot.turnRight((45 - robot.heading).toInt())
            }
        }

        inPosition = true
        waitTurns = 0
        out.println("ANTI-WALLS: En posición esquina '$corner' - Esperando target")
    }

    private fun cornerAmbush() {
        waitTurns++

        if (!targetDetected) {
            robot.turnGunLeft(10)
        }

        if (waitTurns > 50) {
            changeCorner()
        }

        if (waitTurns % 20 == 0) {
            adjustPosition()
        }
    }

    private fun changeCorner() {
        currentCorner = (currentCorner + 1) % 4
        inPosition = false
        targetDetected = false
        out.println("ANTI-WALLS: Cambiando a esquina '$currentCorner'")
    }

    private fun adjustPosition() {
        robot.ahead(20)
        robot.back(10)
    }

    override fun onScannedRobot() {
        val distance: Int = robot.scannedDistance
        val bearing: Int = robot.scannedBearing

        targetDetected = true

        if (distance < 200) {
            robot.turnGunTo(robot.scannedAngle)
            robot.fire(3.0)
            out.println("ANTI-WALLS: ¡Emboscada exitosa! Target a '$distance' px")
        } else if (distance < 400) {
            robot.turnGunTo(robot.scannedAngle)
            robot.fire(2.0)
            out.println("ANTI-WALLS: ¡Disparo de seguimiento a '$distance' px!")
        }

        if (distance < 100) {
            robot.turnRight(robot.scannedBearing + 180)
            robot.ahead(50)
            robot.turnRight(robot.scannedBearing)
        }
    }

    override fun onHitByBullet() {
        out.println("🎯 ANTI-WALLS: ¡Detectado! Cambiando posición")

        robot.turnRight(robot.hitByBulletBearing + 90)
        robot.ahead(100)

        changeCorner()
    }

    override fun onHitWall() {
        robot.back(30)
        robot.turnRight(45)
        out.println("🎯 ANTI-WALLS: Ajuste de pared")
    }

    override fun setRobot(robot: NicoustinRobot) {
        this.robot = robot
    }

    override fun onWin(event: WinEvent) {
        out.println("🏆 ANTI-WALLS: ¡Victoria táctica!")

        // Celebración tipo Walls - movimiento cuadrado
        repeat(4) {
            robot.turnGunRight(90)
            robot.ahead(25)
        }
    }

    override fun evalStrat(): String? {
        return null
    }
}